Use search_after in case/v2 api

gherceg opened the pull request:
## Product Description
<!-- Where applicable, describe user-facing effects and include screenshots. -->

## Technical Summary
<!--
    Provide a link to any tickets, design documents, and/or technical specifications
    associated with this change. Describe the rationale and design decisions.
-->
https://dimagi.atlassian.net/browse/SAAS-17682

### Change in behavior
Due to how `search_after` works, within any given request we know the total number of results that match the underlying query, the page size, and the number of results returned. This means if we have 20 total results, and a page size of 10, we won't know to stop after the second pagination since we aren't tracking how many total elements we've paginated over. This means there will be one additional `next` url returned the user that will return an empty result set.

### Summary

[search_after](https://www.elastic.co/docs/reference/elasticsearch/rest-apis/paginate-search-results#search-after) is the replacement for the scroll API for deep pagination. The case/v2 API has been using range queries to paginate, but performance of that has been poor lately. It isn't entirely clear to us why `search_after` is more performant than range queries, but based on testing (see the ticket) it is definitely the case. My understanding is that `search_after` can take advantage of the ordering of docs (or doc_values) on the segment itself, so that it can efficiently pickup where the last request left off.

As far as how this change impacts users, this code is used when paginating a list of cases in the case/v2 api. We return a `next` url which has a `cursor` parameter that is base64 encoded. Looks like this
```
https://www.commcarehq.org/a/gherceg/api/case/v2?cursor=Y2FzZV90eXBlPWNhc2UmbGltaXQ9MSZpbmRleGVkX29uLmd0ZT0yMDI0LTA1LTMxVDE4JTNBMDQlM0EyMS45MjQ2MDJaJmxhc3RfY2FzZV9pZD04ZjE4MDUzMy1hYjkxLTRhMTItYTgwOC1mMjhiZTViYjVkMGI%3D"
```
When that `cursor` parameter is decoded, it looks something like this 
```
case_type=case&limit=1&indexed_on.gte=b'case_type=case&limit=1&indexed_on.gte=2024-05-31T18%3A04%3A21.924602Z&last_case_id=8f180533-ab91-4a12-a808-f28be5bb5d0b'
&last_case_id=8f180533-ab91-4a12-a808-f28be5bb5d0b
```

Notably, the value of `indexed_on.gte` is what will change with change. So the new decoded parameters will look like
```
case_type=case&limit=1&indexed_on.gte=b'case_type=case&limit=1&indexed_on.gte=1717178661924&last_case_id=8f180533-ab91-4a12-a808-f28be5bb5d0b'
&last_case_id=8f180533-ab91-4a12-a808-f28be5bb5d0b
```

This value is certainly less user friendly as you can't easily determine the date, but the user shouldn't have to interact with this value anyway, since they only need to execute the request with the encoded `cursor` parameter.

In 17170ae447bfbccf93978f9a0ea5f49252acda2c I added backwards compatibility in case anyone triggers a request with a `cursor` parameter that was generated prior to this being deployed. It is a bit messy, but based on my testing does the job. Plus we can remove this pretty shortly after deploying this (a day later?). Here is an example from the shell:
```
In [90]: indexed_on = "2024-06-07T19:11:37.757028Z"
In [91]: timestamp_from_es = 1717787497757
In [92]: timestamp_in_secs = datetime.fromisoformat(indexed_on.replace("Z", "+00:00")).timestamp()
In [93]: timestamp = int(timestamp_in_ms * 1000)
In [94]: timestamp == timestamp_from_es
Out[94]: True
In [95]: timestamp
Out[95]: 1717787497757
```
## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->
- I've confirmed that elasticsearch can handle the timestamp as either a string or int when passed into the query.
- will test on staging

 I've run a lot of es queries from the django shell that use `search_after` so that gives me confidence, though still need to make sure this works as expected end to end.
### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
No

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->
If we revert it, there won't be backwards compatibility for `cursor` parameters that use a timestamp instead of an isoformat date.

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Add search_after helper

Depending on the calling context, it might be easier to support both
passing in a list and individual parameters.

gherceg committed:
Use search_after instead of range queries for pagination

gherceg committed:
Convert isoformat->timestamp for backwards compat

gherceg committed:
Fix issues with search_after helper and caller

gherceg committed:
Paginate until results are empty

gherceg commented:
Thanks to the great pagination test, I discovered an unfortunate side effect with this change which is that `matching_records` will be constant since `search_after` isn't filtering any records out the same way the range filters were. This means the only reliable way to know we've paginated through all results is for the results list to be empty. This was addressed in 0e14e9a0db97896ca8ec9dbea9a8d6dec0cce2d9.

gherceg committed:
Use query._size to determine if more results exist

This also reverts commit 0e14e9a0db97896ca8ec9dbea9a8d6dec0cce2d9.

millerdev reviewed (approved):
I think my comments are non-blocking. I'll let you decide if follow-up is needed.

gherceg commented on corehq/apps/hqcase/tests/test_case_list_api.py:
```diff
@@ -99,7 +99,7 @@ def test_pagination(self):
 
         res = get_list(self.domain, self.couch_user, res['next'])
-        self.assertEqual(res['matching_records'], 2)
+        self.assertEqual(res['matching_records'], 5)
```
Yeah due to my messy commit history I was able to put this clearly in a commit message (my bad). Previously we used range queries which quite literally filtered the number of matching documents down. `search_after` is not applied as a filter, and therefore the matching_records count reflects the total number of docs that match the query.

gherceg committed:
Add test to show paginating until empty result set

This is only necessary when the page size is a divisor of the total
number of records

millerdev commented on corehq/apps/hqcase/tests/test_case_list_api.py:
```diff
@@ -106,6 +106,21 @@ def test_pagination(self):
+        self.assertIn('next', res)
+        res = get_list(self.domain, self.couch_user, res['next'])
+        self.assertFalse(res['cases'])  # empty result set
+        self.assertNotIn('next', res)  # no next param
```
For my curiosity, does the absence of `next` param reset to the first page?

gherceg replied:
The next parameter is in the response, not the request. The next param has a case/v2 url with a `cursor` parameter that contains all of the parameters for **the base request and pagination**, but without the next parameter in the response, there is no next url to execute.

esoergel commented on corehq/apps/hqcase/api/get_list.py:
```diff
@@ -116,10 +116,12 @@ def get_list(domain, couch_user, params):
-    if cases_in_result and es_result.total > cases_in_result:
+    limit = query._size or MAX_PAGE_SIZE
+    if cases_in_result == limit:
+        last_date, last_id = es_result.raw_hits[-1]['sort']
```
There are existing tests for the cursor pagination that hit ES, so I'd expect those would error if this was totally off-base, though there may still be room for more targeted tests, I'm not sure

esoergel replied:
If you can think of a clean way to link the sort parameters to the pagination parameters, that could be useful, though nothing comes to mind.  Tests should prove that it works, though.

esoergel commented on corehq/apps/hqcase/api/get_list.py:
```diff
@@ -116,10 +116,12 @@ def get_list(domain, couch_user, params):
     cases_in_result = len(hits)
-    if cases_in_result and es_result.total > cases_in_result:
+    limit = query._size or MAX_PAGE_SIZE
+    if cases_in_result == limit:
```
I don't follow - won't this still give us a blank page at the end if the number of results is a multiple of page size?

esoergel replied:
Not saying that's a problem, just that the test changes an the commit message made it sound like that would no longer be the case.  Also, why can't this look at the total number of results any more?

esoergel replied:
commented too early - the next commit adds a test that spells it out clearly, though I'm still curious why looking at `result.total` is no longer viable.

gherceg replied:
My understanding is that previously we used range queries which filtered down the total results, but `search_after` instead returns the total number of documents that match the underlying query, and `search_after` just says where to start from. So if there are 11 cases and your page size is 5, the `es_result.total` is going to be 11 every time instead of 5, 5, 1.
